eclipse-theia/theia
# ovsx-client: support multiple registries

Refactors the `OVSXClient` into a simple interface that can easily be implemented by components such as `OVSXHttpClient` or others.

Add a new `OVSXRouterClient` that may route OpenVSX queries to different registries based on an `OVSXRouteConfig` configuration. The path to the configuration file can be specified when starting the backend using the `--ovsx-router-config` CLI param. When this configuration is not specified, we default to the previously configured OpenVSX registry.

The goal of these changes was to minimally impact Theia's plugin code base by simply replacing the client interfaces we use to communicate with an OpenVSX registry. This proved more difficult than expected, especially when merging search results: pagination may currently be broken when dealing with multiple registries.

Add a sample mock OpenVSX server running as part of Theia's backend. You can find the plugins served by this mock server under `examples/api-samples/plugins-db`. This mock is there to allow easier testing of the multi-registry feature.



#### How to test

- Starting the backend without the `--ovsx-router-config` CLI argument should just query `open-vsx.org` for plugins.
- Search extensions for `sample-namespace`: It should only emit a request to the mock OpenVSX server and show its results
- Search for `sample-namespac`: It should emit the request to both (mock, openvsx) registries.

#### Review checklist


#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)
